// @tsed/schema/lib/cjs/components/async-api/payloadMapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.payloadMapper = void 0;
const core_1 = require("@tsed/core");
const change_case_1 = require("change-case");
const JsonParameterTypes_js_1 = require("../../domain/JsonParameterTypes.js");
const SpecTypes_js_1 = require("../../domain/SpecTypes.js");
const JsonSchemaMapperContainer_js_1 = require("../../registries/JsonSchemaMapperContainer.js");
const generics_js_1 = require("../../utils/generics.js");
const somethingOf_js_1 = require("../../utils/somethingOf.js");
function mapOptions(parameter, options = {}) {
    return {
        ...options,
        groups: parameter.groups,
        groupsName: parameter.groupsName
    };
}
function getParameters(jsonOperation, options) {
    return jsonOperation.get("parameters").filter((parameter) => (0, JsonParameterTypes_js_1.isParameterType)(parameter.get("in")));
}
function payloadMapper(jsonOperationStore, operationPath, options) {
    const parameters = getParameters(jsonOperationStore.operation, options);
    const payloadName = (0, change_case_1.pascalCase)([operationPath.path, operationPath.method, "Payload"].join(" "));
    (0, core_1.setValue)(options, `components.schemas.${payloadName}`, {});
    const allOf = parameters
        .map((parameter) => {
        const opts = mapOptions(parameter, options);
        const jsonSchema = (0, JsonSchemaMapperContainer_js_1.execMapper)("item", [parameter.$schema], {
            ...opts,
            ...(0, generics_js_1.popGenerics)(parameter)
        });
        switch (parameter.get("in")) {
            case JsonParameterTypes_js_1.JsonParameterTypes.BODY:
                return jsonSchema;
            case JsonParameterTypes_js_1.JsonParameterTypes.QUERY:
            case JsonParameterTypes_js_1.JsonParameterTypes.PATH:
            case JsonParameterTypes_js_1.JsonParameterTypes.HEADER:
                return {
                    type: "object",
                    properties: {
                        [parameter.get("name")]: jsonSchema
                    }
                };
        }
        return jsonSchema;
    }, {})
        .filter(Boolean);
    return (0, somethingOf_js_1.makeOf)("allOf", allOf);
}
exports.payloadMapper = payloadMapper;
(0, JsonSchemaMapperContainer_js_1.registerJsonSchemaMapper)("payload", payloadMapper, SpecTypes_js_1.SpecTypes.ASYNCAPI);
