// @tsed/schema/lib/cjs/utils/ref.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toRef = exports.createRef = exports.createRefName = void 0;
const core_1 = require("@tsed/core");
const change_case_1 = require("change-case");
const SpecTypes_js_1 = require("../domain/SpecTypes.js");
/**
 * ignore
 * @param options
 */
function getHost(options) {
    const { host = `#/${[SpecTypes_js_1.SpecTypes.OPENAPI, SpecTypes_js_1.SpecTypes.ASYNCAPI].includes(options.specType) ? "components/schemas" : "definitions"}` } = options;
    return host;
}
/**
 * @ignore
 */
function createRefName(name, options) {
    if (options.groups && options.groups.length) {
        return (0, change_case_1.pascalCase)(`${name} ${options.groupsName || options.groups.join(" ")}`);
    }
    return name;
}
exports.createRefName = createRefName;
/**
 * @ignore
 */
function createRef(name, schema, options) {
    const host = getHost(options);
    const ref = {
        $ref: `${host}/${name}`
    };
    const readOnly = schema.isReadOnly;
    const writeOnly = schema.isWriteOnly;
    return (0, core_1.cleanObject)({
        readOnly: readOnly ? true : undefined,
        writeOnly: writeOnly ? true : undefined,
        ...(readOnly || writeOnly
            ? {
                anyOf: [ref]
            }
            : ref)
    });
}
exports.createRef = createRef;
/**
 * @ignore
 */
function toRef(value, schema, options) {
    const name = createRefName(value.getName(), options);
    if (schema) {
        (0, core_1.setValue)(options, `components.schemas.${name}`, schema);
    }
    return createRef(name, value, options);
}
exports.toRef = toRef;
