// @tsed/engines/lib/cjs/components/DustEngine.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DustEngine = void 0;
const tslib_1 = require("tslib");
const cache_js_1 = require("../utils/cache.js");
const path_1 = require("path");
const viewEngine_js_1 = require("../decorators/viewEngine.js");
const Engine_js_1 = require("./Engine.js");
const util_1 = require("util");
let DustEngine = class DustEngine extends Engine_js_1.Engine {
    constructor() {
        super(...arguments);
        this.views = ".";
        this.ext = "dust";
    }
    configure(options) {
        if (options) {
            if (options.ext) {
                this.ext = options.ext;
            }
            if (options.views) {
                this.views = options.views;
            }
            if (options.settings && options.settings.views) {
                this.views = options.settings.views;
            }
        }
        if (!options || (options && !options.cache))
            this.engine.cache = {};
        this.engine.onLoad = async (path, callback) => {
            if ((0, path_1.extname)(path) === "") {
                path += `.${this.ext}`;
            }
            if (path[0] !== "/") {
                path = `${this.views}/${path}`;
            }
            try {
                callback(null, await (0, cache_js_1.read)(path, options));
            }
            catch (er) {
                callback(er);
            }
        };
    }
    $compile(template, options) {
        this.configure(options);
        let templateName;
        if (options.filename) {
            templateName = options.filename.replace(new RegExp("^" + this.views + "/"), "").replace(new RegExp("\\." + this.ext), "");
        }
        return (0, util_1.promisify)(this.engine.compileFn(template, templateName));
    }
};
DustEngine = tslib_1.__decorate([
    (0, viewEngine_js_1.ViewEngine)("dust", {
        requires: ["dustjs-helpers", "dustjs-linkedin"]
    })
], DustEngine);
exports.DustEngine = DustEngine;
