// @tsed/socketio/lib/cjs/services/SocketIOService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SocketIOService = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@tsed/common");
const SocketIO = tslib_1.__importStar(require("socket.io")); // tslint:disable-line: no-unused-variable
const SocketHandlersBuilder_js_1 = require("../class/SocketHandlersBuilder.js");
const SocketProviderMetadata_js_1 = require("../class/SocketProviderMetadata.js");
const io_js_1 = require("../decorators/io.js");
/**
 *
 */
let SocketIOService = class SocketIOService {
    constructor(injector, io) {
        this.injector = injector;
        this.io = io;
        /**
         *
         * @type {Map<any, any>}
         */
        this.namespaces = new Map();
    }
    /**
     *
     * @param {string} namespace
     * @returns {SocketIO.Namespace}
     */
    getNsp(namespace = "/") {
        if (!this.namespaces.has(namespace)) {
            const conf = { nsp: this.io.of(namespace), instances: [] };
            this.namespaces.set(namespace, conf);
            conf.nsp.on("connection", (socket) => {
                conf.instances.forEach((builder) => {
                    builder.onConnection(socket, conf.nsp);
                });
                socket.on("disconnect", (reason) => {
                    conf.instances.forEach((builder) => {
                        builder.onDisconnect(socket, conf.nsp, reason);
                    });
                });
            });
        }
        return this.namespaces.get(namespace);
    }
    /**
     *
     * @param {Provider<any>} provider
     */
    addSocketProvider(provider) {
        const wsConfig = new SocketProviderMetadata_js_1.SocketProviderMetadata(provider.store.get("socketIO"));
        const nspConfig = this.getNsp(wsConfig.namespace);
        const nsps = new Map();
        this.namespaces.forEach((value, nsp) => {
            nsps.set(nsp, value.nsp);
        });
        const builder = new SocketHandlersBuilder_js_1.SocketHandlersBuilder(provider, this.injector).build(nsps);
        nspConfig.instances.push(builder);
    }
};
SocketIOService = tslib_1.__decorate([
    (0, common_1.Service)(),
    tslib_1.__param(1, io_js_1.IO),
    tslib_1.__metadata("design:paramtypes", [common_1.InjectorService, SocketIO.Server])
], SocketIOService);
exports.SocketIOService = SocketIOService;
