// @mui/material/index.js
const React = require('react');
function create(tag) {
  return React.forwardRef(function Component({ children, component: Comp, ...props }, ref) {
    const Tag = Comp || tag;
    return React.createElement(Tag, { ref, ...props }, children);
  });
}
module.exports = {
  Box: create('div'),
  Stack: create('div'),
  List: create('ul'),
  ListItem: create('li'),
  ListItemText: create('span'),
  ListItemButton: create('button'),
  ListItemIcon: create('span'),
  Drawer: create('div'),
  Avatar: create('img'),
  Typography: create('p'),
  Link: create('a'),
  Button: create('button'),
  Badge: create('span'),
  IconButton: create('button'),
  Popover: create('div'),
  Divider: create('hr'),
  Tooltip: create('div'),
  MenuItem: create('div'),
  AppBar: create('div'),
  Toolbar: create('div'),
  InputBase: create('input'),
  ClickAwayListener: create('div'),
  Slide: create('div'),
  Container: create('div'),
  Card: create('div'),
  Dialog: create('div'),
  DialogTitle: create('div'),
  DialogContent: create('div'),
  DialogActions: create('div'),
  Checkbox: create('input'),
  RadioGroup: create('div'),
  Radio: create('input'),
  Grid: create('div'),
  Paper: create('div'),
  CircularProgress: create('div'),
  TextField: create('input'),
  Snackbar: create('div'),
  Alert: create('div'),
  GlobalStyles: create(React.Fragment),
};
