// @tsed/schema/lib/cjs/domain/JsonRequestBody.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonRequestBody = void 0;
const toJsonMapCollection_js_1 = require("../utils/toJsonMapCollection.js");
const JsonMap_js_1 = require("./JsonMap.js");
class JsonRequestBody extends JsonMap_js_1.JsonMap {
    constructor(obj = {}) {
        super(obj);
        this.$kind = "operationRequestBody";
        this.content(obj.content || {});
    }
    description(description) {
        this.set("description", description);
        return this;
    }
    content(content) {
        this.set("content", (0, toJsonMapCollection_js_1.toJsonMapCollection)(content));
        return this;
    }
    addContent(mediaType, schema, examples) {
        const content = this.get("content");
        const mediaContent = new JsonMap_js_1.JsonMap();
        mediaContent.set("schema", schema);
        examples && mediaContent.set("examples", examples);
        content.set(mediaType, mediaContent);
        return this;
    }
    required(required) {
        this.set("required", required);
        return this;
    }
}
exports.JsonRequestBody = JsonRequestBody;
