// @tsed/socketio/lib/cjs/decorators/socketMiddlewareError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SocketMiddlewareError = void 0;
const platform_middlewares_1 = require("@tsed/platform-middlewares");
const core_1 = require("@tsed/core");
const SocketProviderTypes_js_1 = require("../interfaces/SocketProviderTypes.js");
/**
 * Declare a new SocketMiddlewareError.
 *
 * ### Example
 *
 * A middleware can be also used on a `SocketService` either on a class or on a method.
 *
 * Here an example of a middleware:
 *
 * ```typescript
 * import {SocketMiddlewareError, SocketErr, Socket, Args} from "@tsed/socketio";
 *
 * @SocketMiddlewareError()
 * export class ErrorHandlerSocketMiddleware {
 *    async use(@SocketEventName, @SocketErr err: any, @Socket socket: SocketIO.Socket) {
 *        console.error(err);
 *        socket.emit("error", {message: "An error has occured"})
 *    }
 * }
 * ```
 *
 * Then:
 *
 * ```typescript
 * import {SocketService, SocketUseAfter, SocketUseBefore, Emit, Input, Args} from "@tsed/socketio";
 * import {ErrorHandlerSocketMiddleware} from "../middlewares.js";
 * import {User} from "../models/User.js";
 *
 * @SocketService("/my-namespace")
 * @SocketUseAfter(ErrorHandlerSocketMiddleware) // global version
 * export class MySocketService {
 *
 *    @Input("eventName")
 *    @Emit("responseEventName") // or Broadcast or BroadcastOthers
 *    @SocketUseAfter(ErrorHandlerSocketMiddleware)
 *    async myMethod(@Args(0) userName: User) {
 *
 *        console.log(user);
 *        throw new Error("Error");
 *
 *        return user;
 *    }
 * }
 * ```
 *
 * @returns {Function}
 * @decorator
 * @experimental
 */
function SocketMiddlewareError() {
    return (target) => {
        core_1.Store.from(target).merge("socketIO", {
            type: SocketProviderTypes_js_1.SocketProviderTypes.MIDDLEWARE,
            error: true,
            handlers: {
                use: {
                    methodClassName: "use"
                }
            }
        });
        return (0, platform_middlewares_1.Middleware)()(target);
    };
}
exports.SocketMiddlewareError = SocketMiddlewareError;
