// @tsed/socketio/lib/cjs/decorators/socketUseAfter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SocketUseAfter = void 0;
const core_1 = require("@tsed/core");
/**
 * Attach a Socket Middleware to a method or a class.
 *
 * ### Example
 *
 * A middleware can be also used on a `SocketService` either on a class or on a method.
 *
 * Here an example of a middleware:
 *
 * ```typescript
 * import {SocketMiddlewareError, SocketErr, Socket, Args} from "@tsed/socketio";
 *
 * @SocketMiddlewareError()
 * export class ErrorHandlerSocketMiddleware {
 *    async use(@SocketErr() err: any, @Socket socket: SocketIO.Socket) {
 *        console.error(err);
 *        socket.emit("error", {message: "An error has occured"})
 *    }
 * }
 * ```
 *
 * Then:
 *
 * ```typescript
 * import {SocketService, SocketUseAfter, SocketUseBefore, Emit, Input, Args} from "@tsed/socketio";
 * import {ErrorHandlerSocketMiddleware} from "../middlewares.js";
 * import {User} from "../models/User.js";
 *
 * @SocketService("/my-namespace")
 * @SocketUseAfter(ErrorHandlerSocketMiddleware) // global version
 * export class MySocketService {
 *
 *    @Input("eventName")
 *    @Emit("responseEventName") // or Broadcast or BroadcastOthers
 *    @SocketUseAfter(ErrorHandlerSocketMiddleware)
 *    async myMethod(@Args(0) userName: User) {
 *
 *        console.log(user);
 *        throw new Error("Error");
 *
 *        return user;
 *    }
 * }
 * ```
 *
 * @returns {(target: any, propertyKey?: string, descriptor?: PropertyDescriptor) => void}
 * @decorator
 * @experimental
 * @param middlewares
 */
function SocketUseAfter(...middlewares) {
    return (target, propertyKey, descriptor) => {
        if (propertyKey) {
            core_1.Store.from(target).merge("socketIO", {
                handlers: {
                    [propertyKey]: {
                        useAfter: middlewares
                    }
                }
            });
        }
        else {
            core_1.Store.from(target).merge("socketIO", {
                useAfter: middlewares
            });
        }
    };
}
exports.SocketUseAfter = SocketUseAfter;
